rtos/Mutex.h

```diff
@@ -70,17 +70,19 @@ typedef mbed::ScopedLock<Mutex> ScopedMutexLock;
 class Mutex : private mbed::NonCopyable<Mutex> {
 public:
     /** Create and Initialize a Mutex object
-     *
-     * @note You cannot call this function from ISR context.
+     @param recursive A boolean flag used to create a recursive / or not- mutex.
+
+     @note You cannot call this function from ISR context.
     */
-    Mutex();
+    Mutex(bool recursive = true);
```

Comment:
Binary compatibility and code size considerations would make me tend to want to have a new constructor, not add a default parameter. But I can live with this.

If starting from scratch, I might have considered distinct `Mutex` and `RecursiveMutex` types to match C++ `std`, but it's a bit late now, and I'm not convinced the type distinction buys you enough. I don't really see as much value in being explicitly non-recursive as others do.

Can we document the behaviour of calling `lock` twice on a non-recursive mutex? Does it deadlock or assert or MBED_ERROR? I hope the last.

Presumably you're expecting some form of checking behaviour, or you wouldn't need this. A recursive mutex works in every case a non-recursive one does, so the only reason to provide it is to generate that error. There's no performance benefit (in RTX).

Reply:
Thanks @kjbracey-arm . Now added comment for lock() behaviour.

Reply:
After test, we confirm it deadlocks.
If using distinct types, we would have to merge them back anyway in our port. We don't know what is expected from that requirement.

Reply:
If this is a porting thing, where that layer has both types of mutex ,why not implement both that thing's `mutex` and `recursive_mutex` as `Mutex` without modifying it?

That's what we do here: https://github.com/ARMmbed/mbed-os/blob/master/platform/cxxsupport/mstd_mutex

`Mutex` satisfies all non-recursive mutex requirements, unless you have a specific requirement that it deadlock on misuse, which I doubt.

Reply:
Indeed, for this porting layer, why not use `mstd::mutex` and `mstd::recursive_mutex`? Then you can claim you've mapped their types onto distinct types, and the fact those implementations are currently common isn't your business any more.

I bet they use `std::mutex` and `std::recursive_mutex` in other ports, and don't care if they're actually different. (There is no legal way to tell, afaik)

Reply:
In the porting API we have to fill in a single type has to handle both behaviour

Reply:
> handle both behaviour

It has to deadlock? Program correctness requires that the mutex deadlock when misused? The API says that it is defined behaviour to lock a `mutex` you already own, and it causes deadlock? So you have to pass the parameter to `Mutex` here to choose the behaviour? Just using a `Mutex` as-is wouldn't work, because it wouldn't deadlock?

Really?

Or, as I suspect, it has to handle both

1) never being relocked by the same thread
2) sometimes being relocked by the same thread, and working when it happens

Mutex can do both of those as-is.

Reply:
Actually, you could implement any sort of check yourself if you want. You could have

    if (!recursive) {
        MBED_ASSERT(_mutex.get_owner() != ThisThread::get_id());
    }

(or something other than assert) in your wrapper to assert the porting layer precondition. It makes sense for "correctness" testing when writing multiplatform code that you have something to note recursive-or-not misuse if your development platform doesn't care itself, but other platforms implementing the same porting layer do. But that's a "think about when this gets ported to other platforms" thing, rather than a requirement on this platform, so it makes sense to put that requirement up in the porting layer.